### PR TITLE
Find type mapping among interfaces recursively

### LINKED ISSUE #591: Find type mapping among interfaces recursively

### PR DESCRIPTION
**Pull request by nikitin-da**
@artem-zinnatullin PTAL please =)
Closes #591

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
Let's extract type mapping resolving algorithm to the `storio-common`? :)

**Comment by nikitin-da** [comment]
Sure

**Comment by nikitin-da** [comment]
Done

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    @NonNull
+    private final Map<Class<?>, TypeMapping<?>> indirectTypesMappingCache
+            = new ConcurrentHashMap<Class<?>, TypeMapping<?>>();
+
+    public TypeMappingSearcher(@Nullable Map<Class<?>, ? extends TypeMapping<?>> directTypesMapping) {
+        this.directTypesMapping = directTypesMapping;
+    }
+
+    /**
+     * Search algorithm:
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: Let's move description of algorithm to the comment inside of the method, it's not part of documentation :)

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        TypeMapping<T> mapping = getTypeMappingForCurrent(type);
+        if (mapping != null) {
+            return mapping;
+        }
+
+        // Okay, we don't have direct type mapping.
+        // And we don't have cache for indirect type mapping.
+        // Let's find indirect type mapping and cache it!
+
+        Class<?> parentType = type.getSuperclass();
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: `final`

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+     * <p>
+     * Stop search if root parent is Object.class and doesn't implement another interfaces.
+     *
+     * @return direct or indirect type mapping for passed type, or {@code null}.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T> TypeMapping<T> findTypeMappingRecursive(@NonNull final Class<T> type) {
+
+        TypeMapping<T> mapping = getTypeMappingForCurrent(type);
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: Let's make this `final` and create separate refs where needed, I'm afraid it can cause bugs in future after editing

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        // No type mapping found.
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    private <T> TypeMapping<T> getTypeMappingForCurrent(@NonNull final Class<T> type) {
+        if (directTypesMapping == null) {
+            return null;
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: Please add comment: `no known type mappings, looks like user forgot to add them` or something… at first sight I though that it's a bug lol

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            }
+        }
+
+        // No type mapping found.
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    private <T> TypeMapping<T> getTypeMappingForCurrent(@NonNull final Class<T> type) {
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: Let's try to create better name, something like `getDirectOrCachedTypeMapping()`?

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+     * Complexity:
+     * O(n) where n is number of parent types of passed type (pretty fast).
+     * <p>
+     * Stop search if root parent is Object.class and doesn't implement another interfaces.
+     *
+     * @return direct or indirect type mapping for passed type, or {@code null}.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T> TypeMapping<T> findTypeMappingRecursive(@NonNull final Class<T> type) {
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: Not sure I like that public method name contains `recursive` :dancer: 

Maybe just `findTypeMapping()`?

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Common class for search and cache indirect type mapping.
+ */
+public class TypeMappingSearcher {
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: `TypeMappingFinder`?

File: storio-common/src/test/java/com/pushtorefresh/storio/TypeMappingSearcherTest.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+public class TypeMappingSearcherTest {
```
Comment (artem-zinnatullin) [inline storio-common/src/test/java/com/pushtorefresh/storio/TypeMappingSearcherTest.java]: `TypeMappingFinderTest`?

File: storio-common/src/test/java/com/pushtorefresh/storio/TypeMappingSearcherTest.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    interface InterfaceEntity {
+    }
+
+    interface DescendantInterface extends InterfaceEntity {
+    }
+
+    @Test
+    public void shouldReturnNullIfNoTypeMappingsRegistered() {
```
Comment (artem-zinnatullin) [inline storio-common/src/test/java/com/pushtorefresh/storio/TypeMappingSearcherTest.java]: Please also add same test where we pass `null` as `directTypesMapping`

File: storio-common/src/test/java/com/pushtorefresh/storio/TypeMappingSearcherTest.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        //noinspection unchecked
+        TypeMappingSearcher typeMappingSearcher = new TypeMappingSearcher(Collections.EMPTY_MAP);
+
+        assertThat(typeMappingSearcher.findTypeMappingRecursive(ClassEntity.class)).isNull();
+    }
+
+    @Test
+    public void shouldReturnNullIfNotTypeMappingRegisteredForType() {
```
Comment (artem-zinnatullin) [inline storio-common/src/test/java/com/pushtorefresh/storio/TypeMappingSearcherTest.java]: `IfNoType…` instead of `Not`

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java (hunk 6 above)
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java]: Also, it's not recursive at all if type mapping is direct or already cached!

**Comment by nikitin-da** [comment]
fixed=)

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T> TypeMapping<T> findTypeMapping(@NonNull final Class<T> type) {
+
+        // Search algorithm:
+        // Check current class for direct or indirect type mapping from cache.
+        // Walk through all parent classes and interfaces.
+        // If parent type has direct mapping -> we found indirect type mapping!
+        // If current parent type == Object.class or doesn't implement another interfaces ->
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: Looks like we need to change `== Object.class or …` to `== Object.class and`

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        // Stop search if root parent is Object.class and doesn't implement another interfaces.
+
+        final TypeMapping<T> directOrCachedMapping = getDirectOrCachedTypeMapping(type);
+        if (directOrCachedMapping != null) {
+            return directOrCachedMapping;
+        }
+
+        // Okay, we don't have direct type mapping.
+        // And we don't have cache for indirect type mapping.
+        // Let's find indirect type mapping and cache it!
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: `Let's find indirect type mapping recursively`, let's add `recursively` :)

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+        final Class<?> parentType = type.getSuperclass();
+        if (parentType != null && parentType != Object.class) {
+            final TypeMapping<T> mappingOfParent = (TypeMapping<T>) findTypeMapping(parentType);
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: Looks like we won't check interfaces at all until we check all parent classes… Also, we will process parent classes interfaces earlier than child ones, not sure that it's correct. What is your opinion?

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
+package com.pushtorefresh.storio;
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: Let's move it to `storio/internal` package

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMapping.java [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package com.pushtorefresh.storio;
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMapping.java]: `storio.internal` package please

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Common class for search and cache indirect type mapping.
+ */
+public class TypeMappingFinder {
```
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: Let's extract interface from this class and rename implementation to `impl`? Then put interface to public `com.pushtorefresh.storio` class and implementation to `com.pushtorefresh.storio.internal`

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/impl/DefaultStorIOSQLite.java [hunk 17]
```diff
@@ -203,13 +203,8 @@ public DefaultStorIOSQLite build() {
...
         private final Object lock = new Object();
 
-        // Unmodifiable
-        @Nullable
-        private final Map<Class<?>, SQLiteTypeMapping<?>> directTypesMapping;
-
         @NonNull
-        private final Map<Class<?>, SQLiteTypeMapping<?>> indirectTypesMappingCache
-                = new ConcurrentHashMap<Class<?>, SQLiteTypeMapping<?>>();
+        private final TypeMappingFinder typeMappingFinder;
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/impl/DefaultStorIOSQLite.java]: Let's use interface here and pass impl via Builder (with our impl as default)!

File: storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/impl/DefaultStorIOContentResolver.java [hunk 18]
```diff
@@ -193,17 +193,12 @@ public DefaultStorIOContentResolver build() {
 
     protected class LowLevelImpl extends Internal {
 
-        @Nullable
-        private final Map<Class<?>, ContentResolverTypeMapping<?>> directTypesMapping;
-
         @NonNull
-        private final Map<Class<?>, ContentResolverTypeMapping<?>> indirectTypesMappingCache
-                = new ConcurrentHashMap<Class<?>, ContentResolverTypeMapping<?>>();
+        private final TypeMappingFinder typeMappingFinder;
```
Comment (artem-zinnatullin) [inline storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/impl/DefaultStorIOContentResolver.java]: Same as DefaultStorIOSQLite

**Comment by artem-zinnatullin** [comment]
PLUGGABLE TYPE MAPPING!

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java (hunk 13 above)
Comment (nikitin-da) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: I have fixed it. Now it checks first level parent types just for direct type mapping and then begans to search recursively.

**Comment by nikitin-da** [comment]
Plaggable! :electric_plug:

File: storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java (hunk 14 above)
Comment (artem-zinnatullin) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: ok, now it's invalid :)

Comment (nikitin-da) [inline storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java]: Yep, interface is public

**Comment by artem-zinnatullin** [comment]
Please rebase on latest master for code coverage!

**Comment by artem-zinnatullin** [comment]
@nikitin-da codecov patch check shows that only 63% of the new code is covered while we're targeting 80%, can you please try to cover as much as possible?

**Comment by nikitin-da** [comment]
Yep

**Comment by artem-zinnatullin** [comment]
ping @nikitin-da :)

**Comment by nikitin-da** [comment]
I'm in progress. Sorry for delay (

**Comment by nikitin-da** [comment]
@artem-zinnatullin pong =)

**Comment by artem-zinnatullin** [comment]
`SocketTimeoutException`

**Comment by artem-zinnatullin** [comment]
@karlicoss PTAL

File: storio-common/src/test/java/com/pushtorefresh/storio/operations/internal/OnSubscribeExecuteAsBlockingSingleTest.java [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+        verify(preparedOperation).executeAsBlocking();
+        verify(preparedOperation, never()).asRxObservable();
+        verify(preparedOperation, never()).asRxSingle();
+    }
+
+    @SuppressWarnings("CheckResult")
+    @Test
+    public void shouldCallOnErrorIfExceptionOccurred() {
+        //noinspection unchecked
+        final PreparedOperation<Object>  preparedOperation = mock(PreparedOperation.class);
```
Comment (rsinukov) [inline storio-common/src/test/java/com/pushtorefresh/storio/operations/internal/OnSubscribeExecuteAsBlockingSingleTest.java]: empty space =)

Comment (artem-zinnatullin) [inline storio-common/src/test/java/com/pushtorefresh/storio/operations/internal/OnSubscribeExecuteAsBlockingSingleTest.java]: Oh man, please review this PR so we could merge it and include to next
release

**Comment by rsinukov** [comment]
LGTM

**Comment by nikitin-da** [comment]
many thanks, @rsinukov!
